test/hotspot/jtreg/compiler/loopopts/InvariantCodeMotionReassociateCmp.java

```diff
@@ -0,0 +1,315 @@
+/*
+ * Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package compiler.c2.loopopts;
+
+import compiler.lib.ir_framework.*;
+import jdk.test.lib.Asserts;
+
+/*
+ * @test
+ * @bug 8323220
+ * @summary Test loop invariant code motion for cmp nodes through reassociation
+ * @library /test/lib /
+ * @run driver compiler.c2.loopopts.InvariantCodeMotionReassociateCmp
+ */
+public class InvariantCodeMotionReassociateCmp {
+    private static final int size = 500;
```

Comment:
You could even randomize size in a `@Setup` method, and pass it as another argument.

Reply:
If I have a random size input, the loop is actually not a `CountedLoop`, and this transformation only applies to `CountedLoop`. I don't know all the details of `CountedLoop`, but this feels like a miss. Maybe `CountedLoop` can only handle loops with multiple exits of the limit is a constant.

Reply:
Ha, yes, this would be a heavy miss. Can you post a small `Test.java` where such a loop does not get turned into a `CountedLoop`? I'd like to investigate.

Reply:
Might it be that sometimes the size was too small, so that the loop was simply fully unrolled?

Reply:
My mistake. I was reading from a class member variable `size`. Probably cannot be a counted loop because the `size` can change by iteration.

Should be fine if I just pass a `size` variable.

Reply:
If I pass `size` variable, I get some weird matching patterns after strip mining transformations. I avoid this by moving `blackhole()` to before the if condition.